### PR TITLE
Remove `TypeckResults` from `InferCtxt`

### PR DESCRIPTION
**Pull request by camsteffen**
`InferCtxt` currently has `in_progress_typeck_results` which is only used for some diagnostics during typeck. It adds a lifetime which propagates through a lot of code. This PR moves that field into a new helper struct `TypeErrCtxt`.

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -84,6 +84,31 @@ pub use need_type_info::TypeAnnotationNeeded;
 
 pub mod nice_region_error;
 
+/// A helper for building inference-related errors. The `typeck_results`
+/// field is only populated during an in-progress typeck.
+/// Get an instance by calling `InferCtxt::err` or `FnCtxt::infer_err`.
+pub struct InferErrCtxt<'a, 'tcx> {
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:90]: maybe `TypeErrCtxt` and rename `FnCtxt::infer_err` and `InferCtxt::err` to `fn (Infer|FnCtxt)::err_ctxt` :think

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 2]
```diff
@@ -1782,7 +1839,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         self.tcx.note_and_explain_type_err(diag, terr, cause, span, body_owner_def_id.to_def_id());
 
         if let Some(ValuePairs::PolyTraitRefs(exp_found)) = values
-            && let ty::Closure(def_id, _) = exp_found.expected.skip_binder().self_ty().kind()
+            &&
+        let ty::Closure(def_id, _) = exp_found.expected.skip_binder().self_ty().kind()
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:1843]: there are a lot of seemingly unrelated formatting changes here. is there a reason for them?

File: compiler/rustc_typeck/src/check/inherited.rs [hunk 3]
```diff
@@ -124,20 +124,20 @@ impl<'tcx> Inherited<'_, 'tcx> {
 }
 
 impl<'tcx> InheritedBuilder<'tcx> {
-    pub fn enter<F, R>(&mut self, f: F) -> R
+    pub fn enter<F, R>(mut self, f: F) -> R
     where
-        F: for<'a> FnOnce(Inherited<'a, 'tcx>) -> R,
+        F: for<'a> FnOnce(Inherited<'tcx>) -> R,
```
Comment (lcnr) [inline compiler/rustc_typeck/src/check/inherited.rs:129]: `'a` now doesn't exist :grin: though it might be nice to change this function to `&'a FnOnce(&'a Inherited<'tcx>) -> R` so that you still can't use the `Inherited` outside of the closure.

File: compiler/rustc_infer/src/infer/mod.rs [hunk 4]
```diff
@@ -619,20 +603,17 @@ impl<'tcx> InferCtxtBuilder<'tcx> {
         })
     }
 
-    pub fn enter<R>(&mut self, f: impl for<'a> FnOnce(InferCtxt<'a, 'tcx>) -> R) -> R {
+    pub fn enter<R>(&mut self, f: impl for<'a> FnOnce(InferCtxt<'tcx>) -> R) -> R {
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:606]: same here: `'a` is unused :thinking: I think we should either just return an `InferCtxt` here or actually try to change this to `&'a InferCtxt<'tcx>`. Using a closure has the advantage that `enter` can check that `R` does not contain any inference variables afterwards.

Considering that we don't currently check this in `enter`, it seems sensible to just return an `InferCtxt` here.

File: compiler/rustc_infer/src/infer/combine.rs [hunk 5]
```diff
@@ -43,7 +43,7 @@ use rustc_span::{Span, DUMMY_SP};
 
 #[derive(Clone)]
 pub struct CombineFields<'infcx, 'tcx> {
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/combine.rs:45]: can you change all uses of `CombineFields` to `&'a CombineFields<'a, 'tcx>`? don't need 2 separate lifetimes for that

**Review by lcnr — COMMENTED** [review]
really happy about this PR as it is something i've wanted to do myself for quite a while :heart:

some nits and further suggestions, but this looks pretty good already

File: compiler/rustc_infer/src/infer/mod.rs (hunk 4 above)
Comment (lcnr) [reply compiler/rustc_infer/src/infer/mod.rs:606]: cc @rust-lang/types about changing `enter` here

Comment (camsteffen) [reply compiler/rustc_infer/src/infer/mod.rs:606]: Should it be named `build` then? And `build_with_canonical`?

File: compiler/rustc_infer/src/infer/combine.rs (hunk 5 above)
Comment (camsteffen) [reply compiler/rustc_infer/src/infer/combine.rs:45]: I wasn't able to get this to work.

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 1 above)
Comment (camsteffen) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: SGTM

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 2 above)
Comment (camsteffen) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: Should be better now.

File: compiler/rustc_typeck/src/check/inherited.rs (hunk 3 above)
Comment (camsteffen) [reply compiler/rustc_typeck/src/check/inherited.rs]: Removed `for<'a>` and added a reference to the arg.

File: compiler/rustc_infer/src/infer/combine.rs (hunk 5 above)
Comment (lcnr) [reply compiler/rustc_infer/src/infer/combine.rs:45]: ok :thinking: that's surprising :sweat_smile: what are the errors if you do that?

File: compiler/rustc_infer/src/infer/mod.rs (hunk 4 above)
Comment (lcnr) [reply compiler/rustc_infer/src/infer/mod.rs]: jup :+1: renaming this to enter makes sense to me

Comment (camsteffen) [reply compiler/rustc_infer/src/infer/mod.rs]: I'll assume you meant to say `build` 😄

File: compiler/rustc_infer/src/infer/combine.rs (hunk 5 above)
Comment (camsteffen) [reply compiler/rustc_infer/src/infer/combine.rs:45]: I have to use `&'a self` on methods like `fn lub(&'a self, ..) -> Lub<'a, 'tcx>` on `impl CombineFields` and then then after doing that I get `lifetime may not live long enough` where those methods are used.

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 6]
```diff
@@ -54,7 +54,7 @@ struct ConstToPat<'a, 'tcx> {
     behind_reference: Cell<bool>,
 
     // inference context used for checking `T: Structural` bounds.
-    infcx: InferCtxt<'a, 'tcx>,
```
[the remark this answers is not in the document]
Comment (camsteffen) [reply compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:57]: Why not? It was possible since `infer_ctxt().enter()` provided infcx by value.

**Comment by camsteffen** [comment]
I don't know where to start with the failing rustdoc tests 😤

**Comment by lcnr** [comment]
:thinking: i also can't tell why rustdoc is failing :sweat_smile: maybe look at the actually generated docs for the smallest failing test to check whether there are any user-visible changes? :thinking:

**Comment by camsteffen** [comment]
Looks like a bit of waiting and doing nothing made the problem go away!

File: src/librustdoc/clean/blanket_impl.rs [hunk 7]
```diff
@@ -27,76 +27,70 @@ impl<'a, 'tcx> BlanketImplFinder<'a, 'tcx> {
...
+                        let obligation = traits::Obligation::new(
+                            traits::ObligationCause::dummy(),
+                            param_env,
+                            predicate,
+                        );
+                        match infcx.evaluate_obligation(&obligation) {
+                            Ok(eval_result) if eval_result.may_apply() => {}
+                            Err(traits::OverflowError::Canonical) => {}
+                            Err(traits::OverflowError::ErrorReporting) => {}
+                            _ => continue 'blanket_impls,
```
Comment (camsteffen) [inline src/librustdoc/clean/blanket_impl.rs:87]: The issue was here - I was continuing the wrong loop.

File: src/tools/clippy/clippy_lints/src/future_not_send.rs [hunk 8]
```diff
@@ -77,29 +77,29 @@ impl<'tcx> LateLintPass<'tcx> for FutureNotSend {
...
-                                for FulfillmentError { obligation, .. } in send_errors {
-                                    infcx.maybe_note_obligation_cause_for_async_await(db, &obligation);
-                                    if let Trait(trait_pred) = obligation.predicate.kind().skip_binder() {
-                                        db.note(&format!(
-                                            "`{}` doesn't implement `{}`",
-                                            trait_pred.self_ty(),
-                                            trait_pred.trait_ref.print_only_trait_path(),
-                                        ));
-                                    }
+                            let infcx = cx.tcx.infer_ctxt().build();
```
Comment (lcnr) [inline src/tools/clippy/clippy_lints/src/future_not_send.rs:90]: can reuse the `infcx` from above here afaict

**Comment by lcnr** [comment]
small nits which we can fix in a followup pr

prone to merge conflicts:
@bors r+ p=5 rollup=never

File: compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs [hunk 9]
```diff
@@ -63,6 +66,37 @@ pub struct ImplCandidate<'tcx> {
 }
 
 pub trait InferCtxtExt<'tcx> {
```
[the remark this answers is not in the document]
Comment (camsteffen) [reply compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs:68]: Maybe? I was being conservative, only moving stuff to `TypeErrCtxt` if it needs `TypeckResults`.

**Comment by pietroalbini** [comment]
@bors r- retry

Fallout from clicking "synchronize" in the bors UI.

**Comment by lcnr** [comment]
oh wow, should have definitely run perf before approving this 😅 

i guess this resulted in a lot of small inlining/optimization changes, might be worth to check out the flamegraph comparison of a few regressions to check if we can maybe fix them by putting `#[inline]` somewhere.

**Comment by camsteffen** [comment]
Thanks @rylev. I think it might be `traits::overlapping_impls` that needs inline? I see that function in the callgrind diff, and the changes to that function might have made it not inlined anymore.

**Comment by camsteffen** [comment]
Probably [this closure](https://github.com/rust-lang/rust/blob/db0597f5619d5ed93feca28e61226d3581cc7867/compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs#L165-L190), and if `overlapping_impls` is inlined, then the closure would probably be inlined as well?

**Comment by camsteffen** [comment]
`bitmaps` perf addressed in https://github.com/rust-lang/rust/pull/102931. Looks like `deeply-nested-multi` could use another fix.

**Comment by nnethercote** [comment]
`deeply-nested-multi` is less important, plus it has been noisy lately, so don't spend too much time on it.

**Comment by camsteffen** [comment]
Thanks @nnethercote. I'll gladly ignore that.
